core/src/main/java/org/codice/imaging/nitf/core/SharedNitfWriter.java

```diff
@@ -0,0 +1,124 @@
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ */
+package org.codice.imaging.nitf.core;
+
+import java.io.DataOutput;
+import java.io.IOException;
+import java.text.ParseException;
+import org.codice.imaging.nitf.core.dataextension.DataExtensionSegmentWriter;
+import org.codice.imaging.nitf.core.graphic.GraphicSegmentWriter;
+import org.codice.imaging.nitf.core.image.ImageSegmentWriter;
+import org.codice.imaging.nitf.core.label.LabelSegmentWriter;
+import org.codice.imaging.nitf.core.symbol.SymbolSegmentWriter;
+import org.codice.imaging.nitf.core.text.TextSegmentWriter;
+import org.codice.imaging.nitf.core.tre.TreParser;
+
+/**
+ * Output independent parts of a NitfWriter implementation.
+ */
+public abstract class SharedNitfWriter implements NitfWriter {
+
+    private TreParser mTreParser = null;
+    private NitfDataSource mDataSource = null;
+
+    /**
+     * The target to write the data to.
+     */
+    protected DataOutput mOutput = null;
+
+    /**
+     * Initialise shared state.
+     *
+     * @param dataSource the parseStrategy that provides the data to be written out.
+     */
+    protected SharedNitfWriter(final NitfDataSource dataSource) {
+        mDataSource = dataSource;
+    }
+
+    /**
+     * Write out the data to the specified target.
+     *
+     * @throws ParseException if there is a problem reading data
+     * @throws IOException if there is a problem writing data
+     */
+    protected final void writeData() throws ParseException, IOException {
+        mTreParser = new TreParser();
+        NitfFileHeaderWriter fileHeaderWriter = new NitfFileHeaderWriter(mOutput, mTreParser);
+        fileHeaderWriter.writeFileHeader(mDataSource);
+        writeImageSegments();
+        writeGraphicSegments();
+        writeSymbolSegments();
+        writeLabelSegments();
+        writeTextSegments();
+        writeDataExtensionSegments();
+    }
+
+    private void writeImageSegments() throws ParseException, IOException {
+        ImageSegmentWriter imageSegmentWriter = new ImageSegmentWriter(mOutput, mTreParser);
+        int numberOfImageSegments = mDataSource.getImageSegmentHeaders().size();
+        for (int i = 0; i < numberOfImageSegments; ++i) {
+            imageSegmentWriter.writeImageHeader(mDataSource.getImageSegmentHeaders().get(i), mDataSource.getNitfHeader().getFileType());
+            imageSegmentWriter.writeSegmentData(mDataSource.getImageSegmentData().get(i));
+        }
+    }
```

Comment:
If this data were stored in a single list as suggested above, we can write the loops like this:

    for (ImageSegment segment : mDataSource.getImageSegments()) {
      imagSegmentWriter.writeImageHeader(segment.getHeader(), mDataSource.getNitfHeader.getFileType());
      imageSegmentWriter.writeSegmentData(segment.getData());
    }